scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -96,70 +97,109 @@ private object RunnerContext {
 
 /** Convenience object for creating [[ScioContext]] and [[Args]]. */
 object ContextAndArgs {
-  // scalastyle:off regex
-  // scalastyle:off cyclomatic.complexity
-  /** Create [[ScioContext]] and [[Args]] for command line arguments. */
-  def apply(args: Array[String]): (ScioContext, Args) = {
-    val (_opts, _args) = ScioContext.parseArguments[PipelineOptions](args)
-    (new ScioContext(_opts, Nil), _args)
+
+  import scala.language.higherKinds
+  sealed trait ArgsParser[F[_]] {
+    type ArgsType
+    type UsageOrHelp = String
+    type Result = Either[UsageOrHelp, (PipelineOptions, ArgsType)]
+
+    def parse(args: Array[String]): F[Result]
+  }
+
+  final case class DefaultParser[T <: PipelineOptions: ClassTag] private ()
+      extends ArgsParser[Try] {
+    override type ArgsType = Args
+
+    override def parse(args: Array[String]): Try[Result] = Try {
+      Right(ScioContext.parseArguments[T](args))
+    }
   }
 
   import caseapp._
   import caseapp.core.help._
-  def typed[T: Parser: Help](args: Array[String]): (ScioContext, T) = {
-    // limit the options passed to case-app
-    // to options supported in T
-    // fail if there are unsupported options
-    val supportedCustomArgs =
-      Parser[T].args
-        .flatMap { a =>
-          a.name +: a.extraNames
+
+  final case class TypedParser[T: Parser: Help] private () extends ArgsParser[Try] {
+    override type ArgsType = T
+
+    // scalastyle:off regex
+    // scalastyle:off cyclomatic.complexity
+    override def parse(args: Array[String]): Try[Result] = {
+      // limit the options passed to case-app
+      // to options supported in T
+      // fail if there are unsupported options
+      val supportedCustomArgs =
+        Parser[T].args
+          .flatMap { a =>
+            a.name +: a.extraNames
+          }
+          .map(_.name) ++ List("help", "usage")
+
+      val Reg = "^-{1,2}(.+)$".r
+      val (customArgs, remainingArgs) =
+        args.partition {
+          case Reg(a) =>
+            val name = a.takeWhile(_ != '=')
+            supportedCustomArgs.contains(name)
+          case _ => true
         }
-        .map(_.name) ++ List("help", "usage")
-
-    val Reg = "^-{1,2}(.+)$".r
-    val (customArgs, remainingArgs) =
-      args.partition {
-        case Reg(a) =>
-          val name = a.takeWhile(_ != '=')
-          supportedCustomArgs.contains(name)
-        case _ => true
+
+      CaseApp.detailedParseWithHelp[T](customArgs) match {
+        case Left(message) =>
```

Comment:
maybe rename to `case Left(error) => `Failure(new Exception(error.message))` so you don't have to type `message.message` ?